keps/sig-node/4540-strict-cpu-reservation/README.md

```diff
@@ -0,0 +1,465 @@
+
+#### Story 1
+To protect latency of guaranteed workload, systemd daemons including irqbalance daemon are commonly constrained to the reserved CPUs.
+Burstable and best-effort pods (and guaranteed pods with fractional CPU requests) running on the reserved CPUs causes CPU throttling for infrastructure services which results in poor system response time which in turn hits back on workload response time e.g. hand-shaking failures between guaranteed pods.
+
+#### Story 2
+Isolating system daemons and interrupt processing is particularly critical in all-in-one deployments where workloads are placed on combined master+worker(+HCI storage) nodes.
+In all-in-one deployments, system daemons include additional services like ceph storage. Burstable and best-effort pods running on the reserved CPUs can saturate those CPUs degrading infrastructure performance which in turn causes latency issues for workload.
+
+### Risks and Mitigations
+
+The feature is isolated to a specific policy option within the `CPUManager`, and is protected both by the option itself, as well as the `CPUManagerPolicyAlphaOptions` feature gate (which is disabled by default).
+
+Kube-scheduler schedules pods on node allocatable which is total - reserved cores. It is at the node level, burstable and best-effort pods are allowed to run on the reserved cores. The `strict-cpu-reservation` feature removes this discrepancy.
+
+However, kubelet has been requiring non-zero shared pool when the static policy is enabled. Kube-scheduler knows a portion of the node allocatable is not available for exclusive allocation. This not-for-exclusive portion has been conveniently the reserved cores.
+
+To maintain backward compatibility, we introduce `numMinSharedCPUs` in KubeletConfiguration as the minimum number of CPU cores not available for exclusive allocation and expose it to Kube-scheduler.
+
+![MinSharedCPUs](./strict-cpu-allocation.png)
+
+When `strict-cpu-reservation` is disabled:
+```console
+Total CPU cores: 64
+ReservedSystemCPUs: 6
+defaultCPUSet = Reserved (6) + 58 (available for exclusive allocation)
+```
+
+When `strict-cpu-reservation` is enabled:
+```console
+Total CPU cores: 64
+ReservedSystemCPUs: 6
+MinSharedCPUs: 4
+defaultCPUSet = MinSharedCPUs (4) + 54 (available for exclusive allocation)
+```
+
+## Design Details
+
+In Kubelet, when `strict-cpu-reservation` is enabled as a policy option, we remove the reserved cores from the shared pool at the stage of calculation DefaultCPUSet and remove the `MinSharedCPUs` from the list of available cores for exclusive allocation.
+
+Feature impact can be illustrated as following:
+
+With the following Kubelet configuration:
+
+```yaml
+kind: KubeletConfiguration
+apiVersion: kubelet.config.k8s.io/v1beta1
+featureGates:
+  ...
+  CPUManagerPolicyOptions: true
+  CPUManagerPolicyAlphaOptions: true
+cpuManagerPolicy: static
+cpuManagerPolicyOptions:
+  strict-cpu-reservation: "true"
+reservedSystemCPUs: "0,32,1,33,16,48"
+numMinSharedCPUs: 4
```

Comment:
a pretty logical followup would be if we want to let the user specify which cpus, like `reservedSystemCPUs`, or just the count

Reply:
I think just the count is enough since there is no NUMA affinity concern for this parameter.